### PR TITLE
Refactor - Delay visibility of program un-/re-/deployment

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
[Currently some changes to programs made by un-/re-/deployment are visible immediately, even within the same transaction](https://github.com/solana-labs/solana/pull/29725). These inconsistency between top-level instructions and CPI should be removed. This is a preparation for the replacement of the current executor cache by one that supports a [bitemporal model](https://en.wikipedia.org/wiki/Bitemporal_modeling). Once that is done, the new executor cache will further delay the visibility from the end of the transaction to a slot in the future, with a constant offset from the slot the action occurred in.

#### Summary of Changes
- Use three separate `HashMap`s instead of the `enum` `TxBankExecutorCacheDiff`
- Replaces all places which deploy programs by a macro
- Load previous version into the cache before deploying the new one
- Adjusts tests for un-/re-/deployment in same transaction
- Adds a feature gate

#### Old vs New Behavior (inside the same transaction)
- Initial Deployment: Was visible to CPI but not transaction level, will become invisible to both
- Upgrade / Redeployment: Was visible to both, will become unavailable to both
- Close / Undeployment: Was invisible to both, will become visible to both

In other words the new consistent behavior is, that whenever a program is un-/re-/deployed,
it becomes unavailable for the rest of the transaction.

Feature Gate Issue: #30085

### CODE REVIEW

File: program-runtime/src/executor_cache.rs [hunk 1]
```diff
@@ -14,80 +14,69 @@ use {
...
+        executor: Arc<dyn Executor>,
+        upgrade: bool,
+        delay_visibility_of_program_deployment: bool,
+    ) {
+        if upgrade {
+            if !delay_visibility_of_program_deployment {
+                self.executors.insert(key, Some(executor.clone()));
+            } else {
+                // Do not overwrite an existing entry
+                self.executors.entry(key).or_insert(None);
```
Comment (jackcmay) [inline program-runtime/src/executor_cache.rs:58]: Why insert `None` if not found?

Comment (Lichtso) [reply program-runtime/src/executor_cache.rs:58]: That is essentially an explicit invalid entry similar to a tombstone. So that if a program was not deployed before the transaction begins, and it is deployed during the transaction, that the rest of the transaction still sees it as undeployed. In other words it is necessary to get case 1 (deployment invisibility) right.

**Review by taozhu-chicago — COMMENTED** [review]
Catching up on the issue, asked few questions. Will deep dive again.

File: program-runtime/src/executor_cache.rs [hunk 2]
```diff
@@ -14,80 +14,69 @@ use {
...
-
 /// A subset of the BankExecutorCache containing only the executors relevant to one transaction
 ///
 /// The BankExecutorCache can not be updated directly as transaction batches are
 /// processed in parallel, which would cause a race condition.
 #[derive(Default, Debug)]
 pub struct TransactionExecutorCache {
-    pub executors: HashMap<Pubkey, TransactionExecutorCacheEntry>,
+    /// Executors which are visible to the transaction
+    pub executors: HashMap<Pubkey, Option<Arc<dyn Executor>>>,
```
[the remark this answers is not in the document]
Comment (Lichtso) [reply program-runtime/src/executor_cache.rs:24]: executors can contain entries cached from before the transaction began or during the transaction.
And we want to be able to stop new entries from entering during the transaction with the feature gate.
See: https://github.com/solana-labs/solana/pull/29654#discussion_r1098376814

File: program-runtime/src/executor_cache.rs [hunk 3]
```diff
@@ -14,80 +14,69 @@ use {
...
+    /// Executors of programs which were deploymed during the transaction
+    pub deployments: HashMap<Pubkey, Arc<dyn Executor>>,
+    /// Executors which were loaded after the transaction was loaded (because they were missing in the cache)
+    pub loaded: HashMap<Pubkey, Arc<dyn Executor>>,
```
[the remark this answers is not in the document]
Comment (Lichtso) [reply program-runtime/src/executor_cache.rs:28]: No only `executors` is used for visibility, `loaded` is just for updating the banks cache later.

[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply program-runtime/src/executor_cache.rs:28]: Ah, sorry, I didn't word my original comment properly. I meant to ask if executors *added* to `loaded` are meant to be visible immediately since they are also added to `executors` at same time in `set()` function

File: program-runtime/src/executor_cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply program-runtime/src/executor_cache.rs:24]: got it, thx

File: program-runtime/src/executor_cache.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lichtso) [reply program-runtime/src/executor_cache.rs:28]: Only executor which were not loaded before the transaction and were loaded during the transaction, but were **not re-/deployed within the transaction** go into `loaded`.

File: programs/bpf_loader/src/lib.rs [hunk 4]
```diff
@@ -248,11 +254,66 @@ pub fn create_executor_from_account(
...
 
+macro_rules! deploy_program {
+    ($invoke_context:expr, $use_jit:expr, $program_id:expr,
+     $program_account:expr, $programdata_account:expr, $new_programdata:expr,
+     $first_drop:expr, $second_drop:expr $(,)?) => {{
+        let delay_visibility_of_program_deployment = $invoke_context
+            .feature_set
+            .is_active(&delay_visibility_of_program_deployment::id());
+
+        if delay_visibility_of_program_deployment {
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:270]: It's confusing to me that we call `create_executor_from_account` here is `delay_visibility_of_program_deployment` is true, but then `create_executor_from_account` passes false for `delay_visibility_of_program_deployment` to `set()`

File: programs/bpf_loader/src/lib.rs [hunk 5]
```diff
@@ -248,11 +254,66 @@ pub fn create_executor_from_account(
...
+            // Make sure the old version is in the cache so that we don't
+            // load the new version from the database while the old one is still active
+            let mut get_or_create_executor_time = Measure::start("get_or_create_executor_time");
+            let result = create_executor_from_account(
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:274]: deploy might jit compile an program and place it in the cache even if it might never be called by this transaction?

File: runtime/src/bank.rs [hunk 6]
```diff
@@ -4009,22 +4006,18 @@ impl Bank {
...
-            .update_global_cache(&self.executor_cache, |difference| {
-                difference == TxBankExecutorCacheDiff::Inserted
-            });
+            .borrow_mut()
+            .write_back_to_global_cache(&self.executor_cache, false);
     }
 
-    /// Add updated executors back to the bank's cache
+    /// Add upgraded executors back to the bank's cache
     fn store_updated_executors(&self, tx_executor_cache: &RefCell<TransactionExecutorCache>) {
```
Comment (jackcmay) [inline runtime/src/bank.rs:4017]: should this also be renamed to `store_upgraded_executors`?

File: program-runtime/src/executor_cache.rs [hunk 7]
```diff
@@ -14,80 +14,69 @@ use {
...
+        }
     }
 
-    pub fn update_global_cache(
-        &self,
+    pub fn write_back_to_global_cache(
+        &mut self,
         global_cache: &RwLock<BankExecutorCache>,
-        selector: impl Fn(TxBankExecutorCacheDiff) -> bool,
+        deployments: bool,
```
Comment (jackcmay) [inline program-runtime/src/executor_cache.rs:70]: Overloading this function via a bool isn't very clear.  There isn't much overlap, how about making two functions, `write_back_deployments_to_global_cache` and `write_back_loaded_to_global_cache`.  Or maybe better yet a function that returns (via `into_iter()` either deployments or loaded and the caller calls `put()` on the global cache (rather then passing it in)?

File: runtime/src/bank.rs [hunk 8]
```diff
@@ -4009,22 +4006,18 @@ impl Bank {
...
-            .borrow()
-            .update_global_cache(&self.executor_cache, |difference| {
-                difference == TxBankExecutorCacheDiff::Inserted
-            });
+            .borrow_mut()
+            .write_back_to_global_cache(&self.executor_cache, false);
     }
 
-    /// Add updated executors back to the bank's cache
+    /// Add upgraded executors back to the bank's cache
```
Comment (jackcmay) [inline runtime/src/bank.rs:4016]: kinda confusing, upgraded implies redeployed, but I guess also new deployments too?  Maybe `store_deployed_executors` captures both...

File: programs/bpf_loader/src/lib.rs [hunk 9]
```diff
@@ -226,8 +227,13 @@ pub fn create_executor_from_account(
     };
 
     if let Some(ref tx_executor_cache) = tx_executor_cache {
-        if let Some(executor) = tx_executor_cache.get(program.get_key()) {
-            return Ok((executor, None));
+        match tx_executor_cache.get(program.get_key()) {
+            // Executor exists and is cached, use it
+            Some(Some(executor)) => return Ok((executor, None)),
+            // We cached that the Executor does not exist, abort
+            Some(None) => return Err(InstructionError::InvalidAccountData),
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:234]: This looks a little scary since it is a new error conditions that is not directly featurized.  Afaict this case can only occur once the `delay` feature is activated but that is a bit distant and could potentially lead to confusion or an missed code path

File: program-runtime/src/executor_cache.rs (hunk 3 above)
Comment (jackcmay) [inline program-runtime/src/executor_cache.rs:28]: A different names here or a better description would help code readers.

Maybe, `visible`, `deployed`, and `uncached`?

File: programs/bpf_loader/src/lib.rs (hunk 5 above)
Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:274]: Yes, unfortunately that is necessary because we overwrite the original program in its account. So if it was called later there would be no way to load the old version, which is necessary so that the change stays invisible during the transaction.

File: programs/bpf_loader/src/lib.rs (hunk 9 above)
Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:234]: Hm yes, I added a comment to point out that this case only exists when the feature is active.
The thing is, if we made a condition here that would check the feature, it would still result in an error being returned.
Not much gained IMO.

File: programs/bpf_loader/src/lib.rs (hunk 5 above)
Comment (jackcmay) [reply programs/bpf_loader/src/lib.rs:282]: Or we fail any attempts to do so.  Aka, policy is you cannot call a program that you have previously upgraded in the same transaction

File: programs/bpf_loader/src/lib.rs (hunk 9 above)
Comment (jackcmay) [reply programs/bpf_loader/src/lib.rs:237]: Yeah, comment would help

File: programs/bpf_loader/src/lib.rs (hunk 5 above)
Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:282]: Yes that would also be an option. However, this PR is only leading up to the replacement of the executor cache. The new one will extend the delay of visibility of changes way beyond the end of the transaction, to multiple slots.

**Review by jackcmay — DISMISSED** [review]

File: program-runtime/src/executor_cache.rs [hunk 10]
```diff
@@ -9,87 +9,77 @@ use {
...
-            .filter_map(|(key, entry)| {
-                selector(entry.difference).then(|| (key, entry.executor.clone()))
-            })
-            .collect();
-        if !executors_delta.is_empty() {
-            global_cache.write().unwrap().put(&executors_delta);
+        if upgrade {
+            if delay_visibility_of_program_deployment {
+                // Place a tombstone in the cache so that
+                // we don't load the new version from the database as it should remain invisible
```
Comment (taozhu-chicago) [inline program-runtime/src/executor_cache.rs:60]: 👍🏼

**Review by taozhu-chicago — APPROVED** [review]
lgtm, the deploy program macro made tests easier to read too 👍🏼
